### DISCUSSION TITLE
Design Spec for v1.0.0

### DISCUSSION THREAD

**Discussion by dennyabrain** [tattle-made/DAU#60] [Ideas]
# Design Spec
## System Architecture on Staging Environment
![dau-system-arch-staging](https://github.com/tattle-made/DAU_tipline/assets/1415361/3c269de5-d110-48b9-88cc-8178469151c7)

## Contexts
Accounts - Users
UserMessage - Inbox, Outbox
Canon - Manipulated Media, Fact-checked Articles, Analysis
Verification - Query, Comment, Response 

## External Components
|Service | Usecase |
|--- |--- |
|Postgres| Storing data related to dashboard various operations |
|Elasticsearch| Vector and Full Text Search |
|RabbitMQ | Message Broker |
|AWS S3 | Store and Retrieve Media Items |
|AWS SES | Send Emails | 
|Gupshup |BSP offering Whatsapp Business API |

## Worker Nodes at Scale
Incoming media items need to be stored or queried for in our Canon. We have two ways to implement this :
1. Using Oban and Postgres
2. Using RabbitMQ

### 1. Powered by Oban and Postgres
Feluda operators are maintained by feluda team and operators needed for the DAU project (video vector, audio vector) are well tested and have their own specific optimized Dockerfiles available. Feluda Release Manager @duggalsu will be in charge of ensuring regular stable releases of the operators. 

Adding elixir provides the following benefits : 
1. in built clustering - all elixir nodes are connected to each other and help us keep track of overall available resources and helps in allocating jobs equitably without overloading the nodes. This also provides us a way to trigger an auto scale of nodes and pod if needed
2. in built RPCs - without adding any networking library we can communicate with the various worker nodes from the dashboard app. This will also help us get an overall visibility into the worker status and keep track in realtime of successes and failures and catch any issues. 

The elixir layer is mainted in the repo `dau-worker`. Its an elixir app managed using the Mix package manager. Release manager  @dennyabrain will be responsible for ensuring that only the stable builds are pushed to dockerhub; under the tag `stable`. Most recent version of the app can be accessed at `latest`

To prepare a feluda operator to be deployed on kubernetes, we will add an operator specific statement to the worker's dockerfile. For instance to prepare `videovec` operator as a worker we will add `FROM tattletech/feluda-operator-videovec:stable` to the worker specific Dockefile in `dau-worker`. This will make the elixir app along with the elixir environment and the python specific dependencies and operator code and dependencies in the container. Now invoking the command `/app/bin/dau_worker` will make it ready to receive jobs from a Postgres server.

### 2. Powered by RabbitMQ
to be added

**Comment by dennyabrain** [comment]
### Production Infrastructure
![deepfake tech braindump_2024-02-22_10-41-21](https://github.com/tattle-made/DAU_tipline/assets/1415361/6440ddcc-ec07-4971-a3f9-8dfc5b9b7f35)

**Comment by dennyabrain** [comment]
### Incoming Message or Inbox
This controller is part of the UserMessage Context. 

1. Testing Context : `mix test test/dau/user_message_test.exs`
2. Testing Controller : `mix test test/dau_web/controllers/incoming_message_controller_test.exs`

Based on the [API docs](https://docs.gupshup.io/docs/text), I am using the following curl requests to ping the `/gupshup/message` endpoint. This will also be the callback endpoint provided to the BSP.
```shell
curl -H 'Content-Type: application/json' \
-X POST 'http://localhost:4000/gupshup/message' \
-d '{                                                           
"app": "DemoApp",
  "timestamp": 1580227766370,
  "version": 2,
  "type": "message",
  "payload": {
    "id": "ABEGkYaYVSEEAhAL3SLAWwHKeKrt6s3FKB0c",
    "source": "918x98xx21x4",
    "type": "text",
    "payload": {
      "text": "Hi"
    },
    "sender": {
      "phone": "918x98xx21x4",
      "name": "Andy",
      "country_code": "91",
      "dial_code": "8x98xx21x4"
    }
  }
}'
```

**Comment by dennyabrain** [comment]
### Working with Webhooks locally 
Our BSP expects us to configure a webhook URL to which it forwards all messages sent to our whatsapp number. Its not ideal to keep a server 24/7 especially for local development. To address this we are using a webhook payload delivery service [smee](https://smee.io/). Our smee endpoint is `https://smee.io/AFwtodIL1H0jFtW` and this has been configured on our BSP account too. Now when you send any message to your whatsapp number, the smee endpoint receives these message.
> [!CAUTION]
> This url is public so do not send confidential information on it. Remember to periodically delete the messages

Now to receive these events on your local machine, you need to configure smee locally.
1. Install the smee client `npm install --global smee-client`
> [!IMPORTANT]
> Unfortunately this introduces npm dependency on a largely elixir and python project.
2. Run `smee -u https://smee.io/AFwtodIL1H0jFtW -t http://localhost:4000/gupshup/message`
With this process running, all messages from smee are forwarded to your local `dashboard` server.

**Comment by aatmanvaidya** [comment]
The documentation for workers can be found here - https://github.com/tattle-made/feluda/wiki/Workers

**Comment by dennyabrain** [comment]
# Media Matching
User Story : Given a media item, we want to find others like it.
This covers finding duplicates, perceptually similar variants and thematically similar variants of a media item.

The proposed solution will be used to support Spam Filtering and Automatic Response initiatives.

### Definitions
Duplicate : This would mean an exact copy of the media item. Capturing any sort of filters or transformations of the media item is out of scope for this.
Perceptually Similar : This would mean if something a human would count as the same media, so in the case of an image, it would mean the same image with slight resize or crop of it.
Thematically Similar : This would mean that the content of the media item is similar in meaning or context. So maybe all satirical posts are clubbed in together.

The use cases that we need to support are as follows : 
1. Automate as much of the work as possible while allowing for people to override or improve machine labelling.
2. At any time multiple techniques that achieve these labelling could be active. We should be able to switch out the active matching technique without breaking old or new logic
3. Understand that no technique will truly solve for Similarity Matching, hence the scope of this task is to constantly identify shifting definitions of 'matching' and ensure that the database schema and chose techniques can support that.

### Duplicate Matching
Feluda has workers that can process a media item (image, video and audio) and convert them into a hash string. This can be easily stored in a Relational database like Postgres. This makes it an ideal solution to implement optimized match finding and counting instances.

The schema for the Postgres table (user_message_inbox_duplicate) would look like :

| column      | description                                                                                                                                                                              |
| ----------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| id          |               auto generated                                                                                                                                                                           |
| inbox       | references user_message_inbox; type int                                                                                                                                                            |
| value       | computed hash of the media item                                                                                                                                                          |
| worker_name | name of the feluda worker, kept for supporting multiple implementation of hashing techniques. This also allows using different techniques within the dashboard for independent use cases |
| inserted_at |     	timestamp                                                                                                                                                                                     |
|updated_at             |        	timestamp                                                                                                                                                                                  |

Example Use Case : 
1. Given a media item, find its duplicates
   Get the media item's hash value `SELECT value FROM user_message_inbox_duplicate WHERE inbox = $ID`
   Using this obtained value, you can query all items with the same hash as this `SELECT inbox from user_message_inbox_duplicate WHERE value = $VALUE`
   We can also count how many duplicates it has `SELECT COUNT(*) FROM user_message_inbox_duplicate WHERE value = $VALUE` 

### Perceptual Similarity Matching
Feluda has workers that use ML models to convert image, video and audio into vectors. These vector representations of media item allow for finding similar media items and also let us do clustering. With the advancement in multimodal and multilingual models, this technique will continue to be relevant in Feluda. For the use case of perceptual similarity matching, we care to use the vector models simply because they capture information about the media item much better than purely hash based solutions. So to be able to somehow capture that as an identifier and store in a relational database and use for quick querying is worth pursuing. 

The schema for the Postgres table (user_message_inbox_perceptually_similar) would look like :

| column      | description                                                                                                                                                                              |
| ----------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| id          |                                                                                                                                                                                          |
| inbox       | references user_message_inbox; type int                                                                                                                                                            |
| value       | string representation of the vector                                                                                                                                                      |
| worker_name | name of the feluda worker, kept for supporting multiple implementation of hashing techniques. This also allows using different techniques within the dashboard for independent use cases |
| inserted_at |     	timestamp                                                                                                                                                                                     |
|updated_at             |        	timestamp                                                                                                                                                                                  |                                                                                                                                                                               |

Example Use Case : 
1. Given a media item, find its duplicates
   Get the media item's hash value `SELECT value FROM user_message_inbox_perceptually_similar WHERE inbox = $ID`
   Using this obtained value, you can query all items with the same hash as this `SELECT inbox from user_message_inbox_perceptually_similar WHERE value = $VALUE`
   We can also count how many duplicates it has `SELECT COUNT(*) FROM user_message_inbox_perceptually_similar WHERE value = $VALUE` 

#### Alternate Approach
Instead of trying to reduce a vector representation into a string that can be stored in Postgres, I think we could modify the Worker. Consider the scenario when a media item is added to the Queue to be processed. Worker receives the message and generates vector representation (R) of it. Now at this point, we search Elasticsearch for R. If it does not exist, we add it to the index and capture its index id (lets say index_id). If it does exist, we ensure that that the match confidence is (nearly 1.0, signifying exact match) and store its index id (again, index_id). This index_id can then be stored in the reporter queue. This is then picked up by the dashboard backend app and added to the `user_message_inbox_perceptually_similar` in the `value` column. In this scenario the `value` is not really a hash or any sort of representation of the media item. But its just a way to ensure that similar media items have the same id.


### Thematic Similarity Matching
I think an approach similar to the above could be used. Its not urgent to develop this, so we'll revisit this when needed.

**Comment by aatmanvaidya** [comment]
## List of Queue's
@dennyabrain here is the list of queue and its names that will be helpful to you

|  **Worker**  |  **Index Queue**  | **Report Queue (of Index)** |  **Search Queue**  | **Report Queue (of Search)** |
|:------------:|:-----------------:|:---------------------------:|:------------------:|:----------------------------:|
| video worker | video-index-queue |         report-queue        | video-search-queue |      search-result-queue     |
| audio worker | audio-index-queue |         report-queue        | audio-search-queue |      search-result-queue     |
|  hash worker |    count-queue    |      count-report-queue     |          -         |               -              |
| media worker | embedding-index-queue |         embedding-report-queue        |          -         |               -              |

**Comment by dennyabrain** [comment]
I will create two API endpoints
1. /gupshup/sender 
   You send the user phone number in the body of a POST request. I will send a JSON back with all user preference in the database. In the response if the `language` field is nil or empty, this suggests that this is a new user or more importantly that we don't have this user's language preference and hence should show them the language menu. 
2. /gupshup/sender/preference
   This API is used to submit a user's preference to our server. The body of your request should have the language id chosen by the user.
   
### Examples
1. Checking if we have user's language in our database
(untested pseudo code)
```
curl -X POST "https://dau.tattle.co.in/gupshup/sender"
-d '{"sender_number":"919323239392"}'
```
Response when we have user's language in database  :  
200 OK
```
{
	language: "en"
}
```

Response when we have user's language in database  :  
200 OK
```
{}
```

1. Registering the user's language in our database after they have chosen the language
```
curl -X POST "https://dau.tattle.co.in/gupshup/sender/preference"
-d '{"sender_number":"919323239392","language":"te"}'
```
Response 
200 OK

**Comment by dennyabrain** [comment]
Testing all payloads related to BSP : `mix test test/dau/user_message_test.exs`

**Comment by dennyabrain** [comment]
| Sr No | name in sheet                | name in gupshup              | name in code                      | test |
|-------|------------------------------|------------------------------|-----------------------------------|------|
| 1     | deepfake_w_ar_0fc            |                              | deepfake_w_ar_0fc_en              | y    |
| 2     | deepfake_wo_ar_2fc           |                              | deepfake_wo_ar_2fc_en             | y    |
| 3     | manipulated_wo_ar_1fc        |                              | manipulated_wo_ar_1fc_en          | y    |
| 4     | manipulated_wo_ar_2fc        |                              | manipulated_wo_ar_2fc_en          | y    |
| 5     | manipulated_w_ar_0fc         |                              | manipulated_w_ar_0fc_en           | y    |
|       | manipulated_wo_ar_0fc        | manipulated_wo_ar_0fc        | manipulated_wo_ar_0fc_en          | y    |
| 6     | not_manipulated_wo_ar_0fc    |                              | not_manipulated_wo_ar_0fc_en      | y    |
| 7     | not_manipulated_wo_ar_1fc    |                              | not_manipulated_wo_ar_1fc_en      | y    |
| 8     | not_manipualted_wo_ar_2fc    |                              | not_manipualted_wo_ar_2fc_en      | y    |
| 9     | inconclusive_w_ar_0fc        |                              | inconclusive_w_ar_0fc_en          | y    |
| 10    | out_of_scope_english         |                              | out_of_scope_wo_ar_0fc_en         | y    |
| 11    | not_ai_gen_wo_ar_0fc         |                              | not_ai_generated_wo_ar_0fc_en     | y    |
| 12    | not_ai_gen_wo_ar_1fc         |                              | not_ai_generated_wo_ar_1fc_en     | y    |
| 13    | not_ai_gen_wo_ar_2fc         |                              | not_ai_generated_wo_ar_2fc_en     | y    |
| 14    | unsupported_lang             |                              | unsupported_language_wo_ar_0fc_en | y    |
| 15    | ai_gen_wo_ar_0fc             |                              | ai_generated_wo_ar_0fc_en         | y    |
| 16    | ai_gen_w_ar_0fc              |                              | ai_generated_w_ar_0fc_en          | y    |
| 17    | ai_gen_wo_ar_1fc             |                              | ai_generated_wo_ar_1fc_en         | y    |
| 18    | ai_gen_wo_ar_2fc             |                              | ai_generated_wo_ar_2fc_en         | y    |
| 19    | deepfake_w_ar_0fc_hi         | deepfake_w_ar_0fc_hi         | deepfake_w_ar_0fc_hi              | y    |
| 20    | deepfake_wo_ar_1fc_hi        | deepfake_wo_ar_1fc_hi        | deepfake_wo_ar_1fc_hi             | y    |
| 21    | deepfake_wo_ar_2fc_hi        | deepfake_wo_ar_2fc_hi        | deepfake_wo_ar_2fc_hi             | y    |
| 22    | manipulated_w_ar_0fc_hi      | manipulated_w_ar_0fc_hi      | manipulated_w_ar_0fc_hi           | y    |
| 23    | manipulated_wo_ar_1fc_hi     | manipulated_wo_ar_1fc_hi     | manipulated_wo_ar_1fc_hi          | y    |
| 24    | manipulated_wo_ar_2fc_hi     | manipulated_wo_ar_2fc_hi     | manipulated_wo_ar_2fc_hi          | y    |
|       | manipulated_wo_ar_0fc_hi     | manipulated_wo_ar_0fc_hi     | manipulated_wo_ar_0fc_hi          | y    |
| 25    | not_manipulated_wo_ar_0fc_hi | not_manipulated_wo_ar_0fc_hi | not_manipulated_wo_ar_0fc_hi      | y    |
| 26    | not_manipulated_wo_ar_1fc_hi | not_manipulated_wo_ar_1fc_hi | not_manipulated_wo_ar_1fc_hi      | y    |
| 27    | not_manipualted_wo_ar_2fc_hi | not_manipualted_wo_ar_2fc_hi | not_manipualted_wo_ar_2fc_hi      | y    |
| 28    | ai_gen_wo_ar_0fc_hi          | ai_gen_wo_ar_0fc_hi          | ai_generated_wo_ar_0fc_hi         | y    |
| 29    | ai_gen_w_ar_0fc_hi           | ai_gen_w_ar_0fc_hi           | ai_generated_w_ar_0fc_hi          | y    |
| 30    | ai_gen_wo_ar_1fc_hi          | ai_gen_wo_ar_1fc_hi          | ai_generated_wo_ar_1fc_hi         | y    |
| 31    | ai_gen_wo_ar_2fc_hi          | ai_gen_wo_ar_2fc_hi          | ai_generated_wo_ar_2fc_hi         | y    |
| 32    | not_ai_gen_wo_ar_0fc_hi      | not_ai_gen_wo_ar_0fc_hi      | not_ai_generated_wo_ar_0fc_hi     | y    |
| 33    | not_ai_gen_wo_ar_1fc_hi      | not_ai_gen_wo_ar_1fc_hi      | not_ai_generated_wo_ar_1fc_hi     | y    |
| 34    | not_ai_gen_wo_ar_2fc_hi      | not_ai_gen_wo_ar_2fc_hi      | not_ai_generated_wo_ar_2fc_hi     | y    |
| 35    | inconclusive_w_ar_0fc_hi     | inconclusive_w_ar_0fc_hi     | inconclusive_w_ar_0fc_hi          | y    |
| 36    | out_of_scope_hi              | out_of_scope_hi              | out_of_scope_wo_ar_0fc_hi         | y    |
| 37    | unsupported_lang_hi          | unsupported_lang_hi          | unsupported_language_wo_ar_0fc_hi | y    |
| 38    | deepfake_w_ar_0fc_ta         | deepfake_w_ar_0fc_ta         | deepfake_w_ar_0fc_ta              | y    |
| 39    | deepfake_wo_ar_1fc_ta        | deepfake_wo_ar_1fc_ta        | deepfake_wo_ar_1fc_ta             | y    |
| 40    | deepfake_wo_ar_2fc_ta        | deepfake_wo_ar_2fc_ta        | deepfake_wo_ar_2fc_ta             | y    |
| 41    | manipulated_w_ar_0fc_ta      | manipulated_w_ar_0fc_ta      | manipulated_w_ar_0fc_ta           | y    |
| 42    | manipulated_wo_ar_1fc_ta     | manipulated_wo_ar_1fc_ta     | manipulated_wo_ar_1fc_ta          | y    |
| 43    | manipulated_wo_ar_2fc_ta     | manipulated_wo_ar_2fc_ta     | manipulated_wo_ar_2fc_ta          | y    |
|       | manipulated_wo_ar_0fc_ta     | manipulated_wo_ar_0fc_ta     | manipulated_wo_ar_0fc_ta          | y    |
| 44    | not_manipulated_wo_ar_0fc_ta | not_manipulated_wo_ar_0fc_ta | not_manipulated_wo_ar_0fc_ta      | y    |
| 45    | not_manipulated_wo_ar_1fc_ta | not_manipulated_wo_ar_1fc_ta | not_manipulated_wo_ar_1fc_ta      | y    |
| 46    | not_manipualted_wo_ar_2fc_ta | not_manipualted_wo_ar_2fc_ta | not_manipulated_wo_ar_2fc_ta      | y    |
| 47    | ai_gen_wo_ar_0fc_ta          | ai_gen_wo_ar_0fc_ta          | ai_generated_wo_ar_0fc_ta         | y    |
| 48    | ai_gen_w_ar_0fc_ta           | ai_gen_w_ar_0fc_ta           | ai_generated_w_ar_0fc_ta          | y    |
| 49    | ai_gen_wo_ar_1fc_ta          | ai_gen_wo_ar_1fc_ta          | ai_generated_wo_ar_1fc_ta         | y    |
| 50    | ai_gen_wo_ar_2fc_ta          | ai_gen_wo_ar_2fc_ta          | ai_generated_wo_ar_2fc_ta         | y    |
| 51    | not_ai_gen_wo_ar_0fc_ta      | not_ai_gen_wo_ar_0fc_ta      | not_ai_generated_wo_ar_0fc_ta     | y    |
| 52    | not_ai_gen_wo_ar_1fc_ta      | not_ai_gen_wo_ar_1fc_ta      | not_ai_generated_wo_ar_1fc_ta     | y    |
| 53    | not_ai_gen_wo_ar_2fc_ta      | not_ai_gen_wo_ar_2fc_ta      | not_ai_generated_wo_ar_2fc_ta     | y    |
| 54    | inconclusive_w_ar_0fc_ta     | inconclusive_w_ar_0fc_ta     | inconclusive_w_ar_0fc_ta          | y    |
| 55    | out_of_scope_ta              | out_of_scope_tamil           | out_of_scope_wo_ar_0fc_ta         | y    |
| 56    | unsupported_lang_ta          | unsupported_lang_ta          | unsupported_language_wo_ar_0fc_ta | y    |
| 57    | deepfake_w_ar_0fc_te         | deepfake_w_ar_0fc_te         | deepfake_w_ar_0fc_te              | y    |
| 58    | deepfake_wo_ar_1fc_te        | deepfake_wo_ar_1fc_te        | deepfake_wo_ar_1fc_te             | y    |
| 59    | deepfake_wo_ar_2fc_te        | deepfake_wo_ar_2fc_te        | deepfake_wo_ar_2fc_te             | y    |
| 60    | manipulated_w_ar_0fc_te      | manipulated_w_ar_0fc_te      | manipulated_w_ar_0fc_te           | y    |
| 61    | manipulated_wo_ar_1fc_te     | manipulated_wo_ar_1fc_te     | manipulated_wo_ar_1fc_te          | y    |
| 62    | manipulated_wo_ar_2fc_te     | manipulated_wo_ar_2fc_te     | manipulated_wo_ar_2fc_te          | y    |
|       | manipulated_wo_ar_0fc_te     | manipulated_wo_ar_0fc_te     | manipulated_wo_ar_0fc_te          | y    |
| 63    | not_manipulated_wo_ar_0fc_te | not_manipulated_wo_ar_0fc_te | not_manipulated_wo_ar_0fc_te      | y    |
| 64    | not_manipulated_wo_ar_1fc_te | not_manipulated_wo_ar_1fc_te | not_manipulated_wo_ar_1fc_te      | y    |
| 65    | not_manipualted_wo_ar_2fc_te | not_manipualted_wo_ar_2fc_te | not_manipulated_wo_ar_2fc_te      | y    |
| 66    | ai_gen_wo_ar_0fc_te          | ai_gen_wo_ar_0fc_te          | ai_generated_wo_ar_0fc_te         | y    |
| 67    | ai_gen_w_ar_0fc_te           | ai_gen_w_ar_0fc_te           | ai_generated_w_ar_0fc_te          | y    |
| 68    | ai_gen_wo_ar_1fc_te          | ai_gen_wo_ar_1fc_te          | ai_generated_wo_ar_1fc_te         | y    |
| 69    | ai_gen_wo_ar_2fc_te          | ai_gen_wo_ar_2fc_te          | ai_generated_wo_ar_2fc_te         | y    |
| 70    | not_ai_gen_wo_ar_0fc_te      | not_ai_gen_wo_ar_0fc_te      | not_ai_generated_wo_ar_0fc_te     | y    |
| 71    | not_ai_gen_wo_ar_1fc_te      | not_ai_gen_wo_ar_1fc_te      | not_ai_generated_wo_ar_1fc_te     | y    |
| 72    | not_ai_gen_wo_ar_2fc_te      | not_ai_gen_wo_ar_2fc_te      | not_ai_generated_wo_ar_2fc_te     | y    |
| 73    | inconclusive_w_ar_0fc_te     | inconclusive_w_ar_0fc_te     | inconclusive_w_ar_0fc_te          | y    |
| 74    | out_of_scope_te              | out_of_scope_telugu          | out_of_scope_wo_ar_0fc_te         | y    |
| 75    | unsupported_lang_te          | unsupported_lang_te          | unsupported_language_wo_ar_0fc_te | y    |

**Comment by aatmanvaidya** [comment]
above comment is getting very filled with existing templates
making a new one now below - starting from urdu templates.
| sr no | name in sheet                | name in gupshup              | name in code                      | test |
|-------|------------------------------|------------------------------|-----------------------------------|------|
| 1     | deepfake_w_ar_0fc_ur         | deepfake_w_ar_0fc_ur         | deepfake_w_ar_0fc_ur              | y    |
| 2     | deepfake_wo_ar_1fc_ur        | deepfake_wo_ar_1fc_ur        | deepfake_wo_ar_1fc_ur             | y    |
| 3     | deepfake_wo_ar_2fc_ur        | deepfake_wo_ar_2fc_ur        | deepfake_wo_ar_2fc_ur             | y    |
| 4     | cheapfake_wo_ar_0fc_ur       | cheapfake_wo_ar_0fc_ur       | cheapfake_wo_ar_0fc_ur            | y    |
| 5     | cheapfake_w_ar_0fc_ur        | cheapfake_w_ar_0fc_ur        | cheapfake_w_ar_0fc_ur             | y    |
| 6     | cheapfake_wo_ar_1fc_ur       | cheapfake_wo_ar_1fc_ur       | cheapfake_wo_ar_1fc_ur            | y    |
| 7     | cheapfake_wo_ar_2fc_ur       | cheapfake_wo_ar_2fc_ur       | cheapfake_wo_ar_2fc_ur            | y    |
| 8     | manipulated_w_ar_0fc_ur      | manipulated_w_ar_0fc_ur      | manipulated_w_ar_0fc_ur           | y    |
| 9     | manipulated_wo_ar_1fc_ur     | manipulated_wo_ar_1fc_ur     | manipulated_wo_ar_1fc_ur          | y    |
| 10    | manipulated_wo_ar_2fc_ur     | manipulated_wo_ar_2fc_ur     | manipulated_wo_ar_2fc_ur          | y    |
| 11    | manipulated_wo_ar_0fc_ur     | manipulated_wo_ar_0fc_ur     | manipulated_wo_ar_0fc_ur          | y    |
| 12    | not_manipulated_wo_ar_0fc_ur | not_manipulated_wo_ar_0fc_ur | not_manipulated_wo_ar_0fc_ur      | y    |
| 13    | not_manipulated_wo_ar_1fc_ur | not_manipulated_wo_ar_1fc_ur | not_manipulated_wo_ar_1fc_ur      | y    |
| 14    | not_manipulated_wo_ar_2fc_ur | not_manipulated_wo_ar_2fc_ur | not_manipulated_wo_ar_2fc_ur      | y    |
| 15    | ai_gen_wo_ar_0fc_ur          | ai_gen_wo_ar_0fc_ur          | ai_generated_wo_ar_0fc_ur         | y    |
| 16    | ai_gen_w_ar_0fc_ur           | ai_gen_w_ar_0fc_ur           | ai_generated_w_ar_0fc_ur          | y    |
| 17    | ai_gen_wo_ar_1fc_ur          | ai_gen_wo_ar_1fc_ur          | ai_generated_wo_ar_1fc_ur         | y    |
| 18    | ai_gen_wo_ar_2fc_ur          | ai_gen_wo_ar_2fc_ur          | ai_generated_wo_ar_2fc_ur         | y    |
| 19    | not_ai_gen_wo_ar_0fc_ur      | not_ai_gen_wo_ar_0fc_ur      | not_ai_generated_wo_ar_0fc_ur     | y    |
| 20    | not_ai_gen_wo_ar_1fc_ur      | not_ai_gen_wo_ar_1fc_ur      | not_ai_generated_wo_ar_1fc_ur     | y    |
| 21    | not_ai_gen_wo_ar_2fc_ur      | not_ai_gen_wo_ar_2fc_ur      | not_ai_generated_wo_ar_2fc_ur     | y    |
| 22    | inconclusive_w_ar_0fc_ur     | inconclusive_w_ar_0fc_ur     | inconclusive_w_ar_0fc_ur          | y    |
| 23    | out_of_scope_ur              | out_of_scope_ur              | out_of_scope_wo_ar_0fc_ur         | y    |
| 24    | unsupported_lang_ur          | unsupported_lang_ur          | unsupported_language_wo_ar_0fc_ur |   y   |

**Comment by maanasb01** [comment]
Templates Table for Marathi language templates.

| sr. no. | name in gupshup                   | name in codebase                  | tested |
|---------|-----------------------------------|-----------------------------------|------|
|       1 | ai_generated_wo_ar_0fc_mr         | ai_generated_wo_ar_0fc_mr         | y    |
|       2 | ai_generated_w_ar_0fc_mr          | ai_generated_w_ar_0fc_mr          | y    |
|       3 | ai_generated_wo_ar_1fc_mr         | ai_generated_wo_ar_1fc_mr         | y    |
|       4 | ai_generated_wo_ar_2fc_mr         | ai_generated_wo_ar_2fc_mr         | y    |
|       5 | cheapfake_w_ar_0fc_mr             | cheapfake_w_ar_0fc_mr             | y    |
|       6 | cheapfake_wo_ar_0fc_mr            | cheapfake_wo_ar_0fc_mr            | y    |
|       7 | cheapfake_wo_ar_1fc_mr            | cheapfake_wo_ar_1fc_mr            | y    |
|       8 | cheapfake_wo_ar_2fc_mr            | cheapfake_wo_ar_2fc_mr            | y    |
|       9 | deepfake_w_ar_0fc_mar             | deepfake_w_ar_0fc_mr              | y    |
|      10 | deepfake_wo_ar_1fc_mr             | deepfake_wo_ar_1fc_mr             | y    |
|      11 | deepfake_wo_ar_2fc_mr             | deepfake_wo_ar_2fc_mr             | y    |
|      12 | inconclusive_w_ar_0fc_mr          | inconclusive_w_ar_0fc_mr          | y    |
|      13 | manipulated_w_ar_0fc_mr           | manipulated_w_ar_0fc_mr           | y    |
|      14 | manipulated_wo_ar_0fc_mr          | manipulated_wo_ar_0fc_mr          | y    |
|      15 | manipulated_wo_ar_1fc_mr          | manipulated_wo_ar_1fc_mr          | y    |
|      16 | manipulated_wo_ar_2fc_mr          | manipulated_wo_ar_2fc_mr          | y    |
|      17 | not_ai_generated_wo_ar_0fc_mr     | not_ai_generated_wo_ar_0fc_mr     | y    |
|      18 | not_ai_generated_wo_ar_1fc_mr     | not_ai_generated_wo_ar_1fc_mr     | y    |
|      19 | not_ai_generated_wo_ar_2fc_mr     | not_ai_generated_wo_ar_2fc_mr     | y    |
|      20 | not_manipulated_wo_ar_0fc_mr      | not_manipulated_wo_ar_0fc_mr      | y    |
|      21 | not_manipulated_wo_ar_1fc_mr      | not_manipulated_wo_ar_1fc_mr      | y    |
|      22 | not_manipulated_wo_ar_2fc_mr      | not_manipulated_wo_ar_2fc_mr      | y    |
|      23 | out_of_scope_wo_ar_0fc_mr         | out_of_scope_wo_ar_0fc_mr         | y    |
|      24 | unsupported_language_wo_ar_0fc_mr | unsupported_language_wo_ar_0fc_mr | y    |
